src/devices/SoftwareSpi/SoftwareSpi.cs

```diff
@@ -0,0 +1,229 @@
+            //       v       v
+            //       ---------
+            //       |       |
+            // ------         ------
+            //  idle   !idle   idle
+            // note: vertical axis represents idle or !idle (pulse) state and is orthogonal
+            //       to low/high related with GPIO - that part is defined by SPI Mode which
+            //       tells us what GPIO state represents idle and also if the measurement happens
+            //       on pulse start or end
+            if (onPulseEnd)
+            {
+                // When we capture onPulseEnd then we need to start pulse before we send the data
+                // and then trigger the capture on exit
+                _bitTransfer = new ScopeData(
+                    enter: () => {
+                        _controller.Write(_clk, !idle);
+                    },
+                    exit: () => {
+                        controller.Write(_clk, idle);
+                    });
+            }
+            else
+            {
+                _bitTransfer = new ScopeData(
+                    exit: () => {
+                        _controller.Write(_clk, !idle);
+                        _controller.Write(_clk, idle);
+                    });
+            }
+
+            _chipSelect = new ScopeData(
+                enter: () => {
+                    _controller.Write(_cs, _settings.ChipSelectLineActiveState);
+                },
+                exit: () => {
+                    _controller.Write(_cs, !(bool)_settings.ChipSelectLineActiveState);
+                });
+        }
+
+        /// <inheritdoc />
+        public override SpiConnectionSettings ConnectionSettings => _settings;
+
+        /// <inheritdoc />
+        public override void TransferFullDuplex(ReadOnlySpan<byte> dataToWrite, Span<byte> dataToRead)
+        {
+            if (dataToRead.Length != dataToRead.Length)
+            {
+                throw new ArgumentException(nameof(dataToWrite));
+            }
+
+            int bitLen = _settings.DataBitLength;
+            int lastBit = bitLen - 1;
+
+            using (StartChipSelect())
+            {
+                for (int i = 0; i < dataToRead.Length; i++)
+                {
+                    byte readByte = 0;
+                    for (int j = 0; j < bitLen; j++)
+                    {
+                        using (StartBitTransfer())
+                        {
+                            int bit = _settings.DataFlow == DataFlow.MsbFirst ? lastBit - j : j;
+                            bool bitToWrite = ((dataToWrite[i] >> bit) & 1) == 1;
+                            if (ReadWriteBit(bitToWrite))
+                            {
+                                readByte |= (byte)(1 << bit);
+                            }
+                        }
+                    }
+
+                    dataToRead[i] = readByte;
+                }
+            }
+        }
+
+        /// <inheritdoc />
```

Comment:
How does this work since it doesn't override like the others?

Reply:
copy & paste fail :P

Reply:
I wasn't actually sure as I've never used **inheritdoc** before and just curious 😄